src/ucp/core/ucp_am.c

```diff
@@ -469,43 +509,108 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
     return status;
 }
 
+static void ucp_am_zcopy_req_complete(ucp_request_t *req, ucs_status_t status)
+{
+    ucs_assert(req->send.state.uct_comp.count == 0);
+
+    if (req->send.msg_proto.am.reg_desc != NULL) {
+        ucs_mpool_put_inline(req->send.msg_proto.am.reg_desc);
+    }
+
+    ucp_request_send_buffer_dereg(req); /* TODO register+lane change */
+    ucp_request_complete_send(req, status);
+
+}
+
+void ucp_am_zcopy_completion(uct_completion_t *self, ucs_status_t status)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
+    if (req->send.state.dt.offset == req->send.length) {
+        ucp_am_zcopy_req_complete(req, status);
+    } else if (status != UCS_OK) {
+        ucs_assert(req->send.state.uct_comp.count == 0);
+        ucs_assert(status != UCS_INPROGRESS);
+
+        /* NOTE: the request is in pending queue if data was not completely sent,
+         *       just dereg the buffer here and complete request on purge
+         *       pending later.
+         */
+        ucp_request_send_buffer_dereg(req);
+        req->send.state.uct_comp.func = NULL;
+        if (req->send.msg_proto.am.reg_desc != NULL) {
+            ucs_mpool_put_inline(req->send.msg_proto.am.reg_desc);
+        }
```

Comment:
duplication of ucp_am_zcopy_req_complete()
why are we not completing user request here?
what prevents this request from generating additional fragments and using the (already released) memory registration?

Reply:
copied this from tag eager code, will double check

Reply:
maybe need to fix tag eager as wel..